File: server/Commtrac.Api/Services/WorkflowTypeRules.cs
using Commtrac.Api.Models;

namespace Commtrac.Api.Services;

/// <summary>
/// Shared rules for project workflow type selection, legacy WorkflowMode derivation,
/// and resolving a config's effective workflow type id.
/// </summary>
public static class WorkflowTypeRules
{
    public static bool IsInspectionType(WorkflowTypeEntity? type)
    {
        if (type is null) return false;
        return type.Name.Contains("inspection", StringComparison.OrdinalIgnoreCase)
            || type.Id.Contains("inspection", StringComparison.OrdinalIgnoreCase);
    }

    /// <summary>
    /// Derives legacy WorkflowMode from a single catalog workflow type.
    /// Inspection → INSPECTION_ONLY; all other types → INSTALLATION_ONLY (install-style UI).
    /// </summary>
    public static string DeriveWorkflowMode(WorkflowTypeEntity type) =>
        IsInspectionType(type) ? "INSPECTION_ONLY" : "INSTALLATION_ONLY";

    public static bool IsInstallationProjectMode(string workflowMode) =>
        workflowMode is "INSTALLATION_ONLY" or "MIXED";

    /// <summary>Backfill WorkflowTypeId from legacy WorkflowMode (not for MIXED).</summary>
    public static string? DefaultTypeIdForLegacyMode(string? workflowMode) =>
        workflowMode switch
        {
            "INSPECTION_ONLY" => "wftype-inspection",
            "MIXED" => null,
            _ => "wftype-installation",
        };

    public static string? ResolveConfigWorkflowTypeId(
        WorkflowConfigEntity config,
        IEnumerable<WorkflowTypeEntity> types)
    {
        if (!string.IsNullOrWhiteSpace(config.WorkflowTypeId))
            return config.WorkflowTypeId;

        var configType = config.ConfigType?.Trim();
        if (string.IsNullOrEmpty(configType)) return null;

        return types.FirstOrDefault(t =>
                string.Equals(t.Name.Trim(), configType, StringComparison.OrdinalIgnoreCase))
            ?.Id;
    }

    public static bool AssignmentAllowedForProject(
        string? projectWorkflowTypeId,
        string? projectWorkflowMode,
        string effectiveConfigTypeId)
    {
        if (string.IsNullOrWhiteSpace(projectWorkflowTypeId))
        {
            // Legacy MIXED — keep existing behaviour until PM picks a single type.
            return string.Equals(projectWorkflowMode, "MIXED", StringComparison.OrdinalIgnoreCase);
        }

        return string.Equals(projectWorkflowTypeId, effectiveConfigTypeId, StringComparison.Ordinal);
    }
}

File: src/utils/workflowTypeRules.ts
import type { Project, WorkflowMode } from "../types/project";
import type { WorkflowType } from "../types/workflowType";

export function isInspectionWorkflowType(type?: Pick<WorkflowType, "id" | "name"> | null): boolean {
  if (!type) return false;
  const id = type.id.toLowerCase();
  const name = type.name.toLowerCase();
  return id.includes("inspection") || name.includes("inspection");
}

/** Derives legacy WorkflowMode from a single catalog workflow type. */
export function deriveWorkflowMode(type: Pick<WorkflowType, "id" | "name">): WorkflowMode {
  return isInspectionWorkflowType(type) ? "INSPECTION_ONLY" : "INSTALLATION_ONLY";
}

export function isInstallationProjectMode(workflowMode?: WorkflowMode | null): boolean {
  return workflowMode === "INSTALLATION_ONLY" || workflowMode === "MIXED" || !workflowMode;
}

/** Backfill WorkflowTypeId from legacy WorkflowMode (not for MIXED). */
export function defaultTypeIdForLegacyMode(workflowMode?: WorkflowMode | null): string | null {
  switch (workflowMode) {
    case "INSPECTION_ONLY":
      return "wftype-inspection";
    case "MIXED":
      return null;
    default:
      return "wftype-installation";
  }
}

/** Resolve the form/catalog type id from a project row (post-migration backfill aware). */
export function resolveProjectWorkflowTypeId(
  project: Pick<Project, "workflowTypeId" | "workflowMode" | "isInstallationProject">,
): string {
  if (project.workflowTypeId) return project.workflowTypeId;
  const legacyMode =
    project.workflowMode ?? (project.isInstallationProject ? "INSTALLATION_ONLY" : "INSPECTION_ONLY");
  return defaultTypeIdForLegacyMode(legacyMode) ?? "";
}

export function findWorkflowType(types: WorkflowType[], id?: string | null): WorkflowType | undefined {
  if (!id) return undefined;
  return types.find((type) => type.id === id);
}

/** True when legacy MIXED project has not yet been assigned a single catalog type. */
export function isLegacyMixedProjectWithoutType(
  project: Pick<Project, "workflowTypeId" | "workflowMode"> | null | undefined,
): boolean {
  if (!project) return false;
  return !project.workflowTypeId && project.workflowMode === "MIXED";
}

/** Whether a config's effective type may be assigned on this project (mirrors server guard). */
export function assignmentAllowedForProject(
  projectWorkflowTypeId: string | null | undefined,
  projectWorkflowMode: WorkflowMode | null | undefined,
  effectiveConfigTypeId: string,
): boolean {
  if (!projectWorkflowTypeId?.trim()) {
    return projectWorkflowMode === "MIXED";
  }
  return projectWorkflowTypeId === effectiveConfigTypeId;
}

/** True when the project restricts assignments to one catalog type (not legacy MIXED). */
export function projectScopesWorkflowType(
  project: Pick<Project, "workflowTypeId" | "workflowMode" | "isInstallationProject"> | null | undefined,
): boolean {
  if (!project) return false;
  return Boolean(resolveProjectWorkflowTypeId(project));
}
